### DISCUSSION TITLE
[Feature Request]: Event counter for rain or flow gauges

### DISCUSSION THREAD

**Discussion by seagull9000** [meshtastic/firmware#5572] [Ideas]
### Platform

other

### Description

Some sensors may trigger a burst of activity that would not be friendly to the mesh.  For example, a rain sensor may trigger rapidly during a torrential downpour  The minimum reporting interval would mitigate this impact, but with the current functionality, valuable information would be lost.

Proposed enhancement:
1.  Should have: Accumulate a count of sensor activations between reports.  This could be a lifetime count (not reset after each report) so that missed reports will not compromise the integrity of the data (client application can just subtract current received from previous received).
2.  Nice to have: Attribute a quantitative value for each activation.  For example, for a rain gauge, the amount in millilitres per activation could be entered and accummulated as in 1.

The CLI/API could have a function to reset accummulated values to 0.  This would provide flexibility so that client applications don't have to worry about starting values.

**Comment by garthvh** [comment]
There is already an interval for how frequently the message is sent out, this is how you reduce mesh traffic, it also no longer works on the default channel if you upgrade your firmware.

**Comment by seagull9000** [comment]
> There is already an interval for how frequently the message is sent out, this is how you reduce mesh traffic, it also no longer works on the default channel if you upgrade your firmware.

Yes, thanks, but the issue is with the loss of data if multiple detections occur between the minimum interval.  I tried to find the relevant protobuf to see if a count already exists, but not sure which it is.

The key point is that messages can be missed, so if the count is important, we can't rely on the client application to keep track of it if the message is lost.

**Comment by seagull9000** [comment]
Unless I have completely missed something here, the current functionality would mean that a repeating event could not be relied upon, so additional hardware would need to be implemented to keep track of the count.

So I don't understand why this is closed without discussion.

**Reply by fifieldt** [reply]
@seagull9000 - converted this to a discussion ... so discussion can continue :)

**Reply by GUVWAF** [reply]
I do see the value of a counter for the number of detections within the transmission interval.

**Comment by seagull9000** [comment]
The feature would support any measuring devices with pulsed output such as energy meters and flow meters, just to add a couple more to the use case.

If implementing a counter breaks the Detection Sensor module, then perhaps a new Event Counter module could be on the cards?

**Reply by GUVWAF** [reply]
The DetectionSensor module is intended to be a very simple module, using text messages. Parsing the text message with the counter is likely not desirable.
The [RemoteHardware module](https://meshtastic.org/docs/configuration/module/remote-hardware/) is more advanced and probably more suited for things like this, but it's not so easy in use.
I'm not sure if yet another module is the way to go, but the DetectionSensor module should be kept simple.

**Reply by seagull9000** [reply]
I see what you mean.

What if we keep the DetectionSensor Module much the way it is, but add an internal counter that can be polled from the client application?  That way protobufs and set up configuration won't need to change.

My thinking is that a GPIO detection occurs, a message is sent, as happens now, but increments a lifetime* counter. held on the node.
The client application is in control of polling the node for the counter value at its own interval.

The benefit of this approach is that missed messages will not result in unreliable/unusable data.

I'm not familiar enough to know what mechanism/message format should be used for polling, but an existing method would be preferred.  We just need the count.

*If maintaining a counter between restarts / power cycles is impractical, then setting to zero on boot would be fine, I think.  The application could test if the counter is less than the previous value and manage that accordingly.  That may result in some edge cases where some data is lost, I don't think the aim should be to have absolutely perfect data - just meaningful data.

Thoughts?

**Reply by GUVWAF** [reply]
Polling would need its own protobuf message to indicate that you are specifically asking for a DetectionSensor counter report, and then the report itself should probably be its own protobuf message. 

@thebentern what do you think about a general `PROTO` mode for the DetectionSensor module that allows for things like this? Then, instead of sending a text message upon detection or at the status update, it could be a dedicated protobuf message containing a better parsable status indicator and e.g. also this counter value?

**Reply by alexschultze** [reply]
+1 for this request. 

My recommendation is to implement a telemetry / environment sensor, which either sets up a counter (e.g., PCNT) or less optimal an interrupt generated incrementing counter on a GPIO. 

Telemetry would be send out in the specified intervals only.

**Comment by seagull9000** [comment]
I've changed the title of this discussion to reflect what has been discussed here.  I still believe there is a need for this feature.

**Comment by maxlock** [comment]
Hi I'd like to add another use case, I have a vibration sensor that closes it's contacts at a frequency proportional to it's movement, or none if no movement. I'm planning to use it as a tamper alarm. A possible feature for the module may be a Threshold setting that alerts only if more than n events in a time period are exceeded.

**Reply by seagull9000** [reply]
What would you expect the upper frequency to be for the vibration sensor?

**Reply by maxlock** [reply]
Thats a very good question, I'll have to do a little testing with the sensor (an SW-420) I appreciate that the gpio is most likely being polled rather than having an interrupt attached which is probably where this discussion is going :) I think I may need resort to external circuitry to realise threshold controls.

**Comment by alexschultze** [comment]
An intermediate solution can be implemented presumably by a custom I2C sensor,  https://github.com/meshtastic/i2c-sensor .

A quite simple missing link would be to include telemetry fields. #10518 is concerning with this.

**Comment by evdmaaten** [comment]
I would strongly support adding this feature.
My use case is a tipping-bucket rain gauge connected by a reed switch to a GPIO pin on a Meshtastic node. Each bucket tip produces a pulse. The node should count and debounce these pulses locally, convert the count into rainfall using a configurable value such as 0.2 mm per pulse, and transmit the accumulated result only at a configurable interval.
Sending a Meshtastic message for every pulse could overload the mesh during heavy rainfall. It also makes the rainfall total unreliable if an individual packet is lost.
Ideally, an Event Counter or Telemetry function would provide:
- Configurable GPIO input and pull-up/pull-down.
- Configurable debounce time.
- A cumulative pulse counter.
- A configurable multiplier or calibration value per pulse.
- Periodic telemetry containing the cumulative count and computed value.
- Counter persistence across restarts, if practical.
- A way to reset or preset the counter.
The cumulative counter is important because the receiving application can calculate the difference between reports. If one telemetry packet is lost, the next report still contains the correct total.
This would also support flow meters, energy meters, anemometers and other sensors with pulse outputs.
My current application is:
reed switch → Meshtastic GPIO → debounced counter → rainfall calculation → periodic telemetry
Would the Meshtastic team consider implementing this as a general GPIO Event Counter in the Telemetry system?

**Reply by evdmaaten** [reply]
At present, I have to use a separate Raspberry Pi Pico to debounce the reed-switch pulses, maintain the counter, calculate the rainfall in millimetres and periodically pass the result to the Meshtastic node. This works, but it adds another microcontroller, additional wiring, power consumption and another possible point of failure.
If Meshtastic could perform this relatively simple calculation directly on the node, the Pico would no longer be necessary. It would make compact, low-power rain gauges and other pulse-based sensors much easier to build.
